packages/flutter/lib/src/widgets/routes.dart

```diff
@@ -618,6 +626,56 @@ abstract class ModalRoute<T> extends TransitionRoute<T> with LocalHistoryRoute<T
   Animation<double> get forwardAnimation => _forwardAnimationProxy;
   ProxyAnimation _forwardAnimationProxy;
 
+  /// If this route [isCurrent] then return the value of the
+  /// [scopedWillPopCallback]. Otherwise return false.
+  ///
+  /// Typically this method is not overridden because applications usually
+  /// don't create modal routes directly, they use higher level primitives
+  /// like [showDialog].  The [scopedWillPopCallback] makes it possible for a
+  /// ModalRoute's descendant to define the value of `willPop`.
+  ///
+  /// See also:
+  ///
+  /// * [Form], which provides an `onWillPop` callback that uses this mechanism.
+  @override
+  Future<bool> willPop() async {
+    if (!isCurrent)
+      return new Future<bool>.value(false);
+    assert(_scopeKey?.currentState != null);
+    final WillPopCallback callback = _scopeKey.currentState.willPopCallback;
+    return callback != null ? callback() : new Future<bool>.value(true);
+  }
+
+  /// Enables this route to veto attempts by the user to dismiss it. This
+  /// property may only be set when this route [isCurrent].
```

Comment:
The _ModalScope is just another widget. If you're in the tree in a Navigator in a ModalRoute, then you must have a _ModalScope because the ModalRoute builds it.